### DISCUSSION TITLE
`PatternBuilder` API tweaks

### DISCUSSION THREAD

**Discussion by skeate** [jacob-alford/schemata-ts#120] [General]
What's available inside `PatternBuilder` right now is pretty barebones; it has enough to do what it has to do, but it feels pretty clunky manually writing out every little step.

Fortunately, much like the whole `Schemable` system, it's composable! So, for example, while it doesn't (right now) have a `digit` combinator, it's pretty easy to write one:

```ts
export const digit = PB.characterClass(false, ['0', '9'])
```

Also, some of the names make sense in some contexts but aren't exactly what I'd pick in others. As an example, when I was writing the docs-ts example for the module, I did things like use `PB.exactly(3)(foo)` instead of `PB.times(3)`; exactly makes sense (to me) in that style, while the `times` naming was chosen thinking about piping (e.g. `pipe(foo, PB.times(3))`)

Just wanted to start a thread to get discussion on

- What helpers (like `digit`) should the PB module export?
- What are some better names or aliases of the existing functions?

**Comment by rae-hill** [comment]
I'm all for adding aliases for common enough helpers, some for `/[A-Z]/` or `/[a-z]/` or `/[A-z]/` would be nice, maybe alpha with a `caps: bool` option.  I'm going to look through the PatternBuilder module to see if I can think of any suggestions

**Comment by skeate** [comment]
I think, at a minimum, anything that exists as a regex character class (e.g. `\d` which is digit, `\w` "word character", etc). there are also the [POSIX named character classes](https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions#Character_classes) (which have some overlap)

**Reply by rae-hill** [reply]
So this would basically be `upper`, `lower`, `alpha`, `digit`, ..., `space`, ...etc?

**Comment by skeate** [comment]
Oh, another idea: number range helpers. e.g. `range(20, 55)` expands to something like

```ts
pipe(
  PB.characterClass(false, ['2', '4']),
  PB.then(digit),
  PB.or(pipe(
    PB.char('5'),
    PB.then(PB.characterClass(false, ['0', '5'])),
  )),
)
```
wouldn't be fun to write but would be immensely helpful

**Reply by rae-hill** [reply]
That would be super helpful!
